finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala

```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
   def += (kv: (String, Date)): HeaderMap =
     += ((kv._1, HeaderMap.format(kv._2)))
 
-  override def empty: HeaderMap = new MapHeaderMap(mutable.Map.empty)
+  override def empty: HeaderMap = MapHeaderMap()
 }
 
+private[finagle] case class HeaderValuePair(header: String, value: String) {
+  val canonicalName = HeaderValuePair.canonicalName(header)
+}
+
+private[finagle] object HeaderValuePair {
+  def canonicalName(header: String) = header.toLowerCase
+}
 
 /** Mutable-Map-backed [[HeaderMap]] */
-class MapHeaderMap(underlying: mutable.Map[String, Seq[String]]) extends HeaderMap {
+class MapHeaderMap extends HeaderMap {
+
+  private val underlying = mutable.Map.empty[String, List[HeaderValuePair]]
+
+  // This is here to provide backward compatibility
+  def this(underlying: mutable.Map[String, Seq[String]]) {
```

Comment:
You have a valid point here. I was going to suggest adding a deprecation here. But I personally prefer breaking the api by removing this constructor. Reasons:
* there is already an apply method which takes tuples. I would expect that most people are using that over this constructor.
* this is a pretty easy fix on the users' side. I don't think it worths the trouble of deprecating it.

Let me know which way you want to go with. Happy to do the changes.